### PR TITLE
As/add account manager support

### PR DESCRIPTION
**Pull request by adamskrodzki**

### CODE REVIEW

File: contracts/adapters/MakerAdapter.sol [hunk 1]
```diff
@@ -24,15 +24,21 @@ import "../interfaces/MPALike.sol";
...
 import "../McdView.sol";
 import "../McdUtils.sol";
-import "hardhat/console.sol";
 
 contract MakerAdapter {
     ServiceRegistry public immutable serviceRegistry;
     address private immutable dai;
     string private constant CDP_MANAGER_KEY = "CDP_MANAGER";
     string private constant MCD_UTILS_KEY = "MCD_UTILS";
+    address public immutable self;
```
Comment (halaprix) [inline contracts/adapters/MakerAdapter.sol:33]:
```suggestion
        address private immutable self;
```

File: contracts/adapters/MakerAdapter.sol [hunk 2]
```diff
@@ -69,6 +75,10 @@ contract MakerAdapter {
         ManagerLike manager = ManagerLike(serviceRegistry.getRegisteredService(CDP_MANAGER_KEY));
         (uint256 cdpId, ) = decode(triggerData);
         address cdpOwner = manager.owns(cdpId);
+        require(
+            canCall(address(this), manager, cdpId, cdpOwner),
+            "maker-adapter/not-allowed-to-call"
+        ); //missing check to fail permit if msg.sender has no permissions
```
Comment (halaprix) [inline contracts/adapters/MakerAdapter.sol:81]: we might want to switch ti custom errors in the entire repository - if time allows.

File: contracts/adapters/DPMAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import "../interfaces/MPALike.sol";
+import "../ServiceRegistry.sol";
+import "../McdView.sol";
+import "../McdUtils.sol";
+
+contract DPMAdapter {
+    ServiceRegistry public immutable serviceRegistry;
+    string private constant CDP_MANAGER_KEY = "CDP_MANAGER";
+    string private constant MCD_UTILS_KEY = "MCD_UTILS";
+    address public immutable self;
```
Comment (halaprix) [inline contracts/adapters/DPMAdapter.sol:33]:
```proposal
    address private immutable self;
```

File: contracts/AutomationBot.sol [hunk 4]
```diff
@@ -307,7 +318,8 @@ contract AutomationBot {
         ICommand command = ICommand(commandAddress);
 
         require(command.isExecutionLegal(triggerData), "bot/trigger-execution-illegal");
-        IAdapter adapter = IAdapter(getAdapterAddress(commandAddress));
+        IAdapter dpmAdapter = IAdapter(getAdapterAddress(commandAddress, false));
+        IAdapter adapter = IAdapter(getAdapterAddress(commandAddress, true));
```
Comment (halaprix) [inline contracts/AutomationBot.sol:322]: to be consistent with other nonexecutable adapters being called `adapter`
```proposal
        IAdapter adapter = IAdapter(getAdapterAddress(commandAddress, false));
        IAdapter executableAdapter = IAdapter(getAdapterAddress(commandAddress, true));
```

File: contracts/AutomationBot.sol [hunk 5]
```diff
@@ -317,14 +329,19 @@ contract AutomationBot {
...
             )
         );
-        require(status, "bot/failed-to-draw-dai");
+        require(status, "bot/failed-to-draw-coverage");
         {
-            (bool statusAllow, ) = address(adapter).delegatecall(
-                abi.encodeWithSelector(adapter.permit.selector, triggerData, commandAddress, true)
+            (bool statusAllow, ) = address(dpmAdapter).delegatecall(
+                abi.encodeWithSelector(
+                    dpmAdapter.permit.selector,
```
Comment (halaprix) [inline contracts/AutomationBot.sol:336]: same comment
```proposal
            (bool statusAllow, ) = address(adapter).delegatecall(
                abi.encodeWithSelector(
                    adapter.permit.selector,
```

File: contracts/AutomationBot.sol [hunk 6]
```diff
@@ -307,7 +318,8 @@ contract AutomationBot {
         ICommand command = ICommand(commandAddress);
 
         require(command.isExecutionLegal(triggerData), "bot/trigger-execution-illegal");
-        IAdapter adapter = IAdapter(getAdapterAddress(commandAddress));
+        IAdapter dpmAdapter = IAdapter(getAdapterAddress(commandAddress, false));
+        IAdapter adapter = IAdapter(getAdapterAddress(commandAddress, true));
         (bool status, ) = address(adapter).delegatecall(
             abi.encodeWithSelector(
                 adapter.getCoverage.selector,
```
Comment (halaprix) [inline contracts/AutomationBot.sol:325]:
```suggestion
        (bool status, ) = address(executableAdapter).delegatecall(
            abi.encodeWithSelector(
                executableAdapter.getCoverage.selector,
                ```

File: contracts/AutomationBot.sol [hunk 7]
```diff
@@ -317,14 +329,19 @@ contract AutomationBot {
...
+                    true
+                )
             );
             require(statusAllow, "bot/permit-failed");
 
-            command.execute(executionData, triggerData);
+            command.execute(executionData, triggerData); //command must be whitelisted
             (, , bool continuous) = automationBotStorage.activeTriggers(triggerId);
             if (!continuous) {
                 clearTrigger(triggerId);
```
Comment (halaprix) [inline contracts/AutomationBot.sol:347]: lines 350 - 352 :
```diff
-             (bool statusDisallow, ) = address(adapter).delegatecall(
-                abi.encodeWithSelector(adapter.permit.selector, triggerData, commandAddress, false)
-            );
+             (bool statusDisallow, ) = address(executableAdapter).delegatecall(
+                abi.encodeWithSelector(executableAdapter.permit.selector, triggerData, commandAddress, false)
+            );
```

File: scripts/tasks/open-aave-position.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        /*
+        const creationReceipt2 = await (
+            await account.execute(aave_pa.address, encodedDrawDebtData, {
+                value: EthersBN.from(args.eth.toString()).mul(EthersBN.from(10).pow(18)),
+                gasLimit: 3000000,
+            })
+        ).wait()
+
+        console.log('drawDebt', creationReceipt2)
+*/
```
Comment (halaprix) [inline scripts/tasks/open-aave-position.ts:61]: should that work or should we remove it or we'll handle it later ?

File: scripts/tasks/add-dpmproxy-trigger.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        console.log('adding trigger', creationReceipt)
+        /*
+        const creationReceipt2 = await (
+            await account.execute(aave_pa.address, encodedDrawDebtData, {
+                value: EthersBN.from(args.eth.toString()).mul(EthersBN.from(10).pow(18)),
+                gasLimit: 3000000,
+            })
+        ).wait()
+
+        console.log('drawDebt', creationReceipt2)
```
Comment (halaprix) [inline scripts/tasks/add-dpmproxy-trigger.ts:64]: do we TODO it or remove it ?

**Review by halaprix — APPROVED** [review]

File: scripts/tasks/open-aave-position.ts (hunk 8 above)
Comment (adamskrodzki) [reply scripts/tasks/open-aave-position.ts:61]: can be removed

**Pull request merged**